proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,483 @@
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+Grammar
+~~~~~~~
+
+Add a new ``-XStringInterpolation`` extension, which, when enabled, adds a new expression to `Chapter 3 of the Haskell 2010 Report <https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-220003>`_ after do-expressions in the ``exp`` rule:
+
+.. code-block:: abnf
+
+  exp → ...
+      | case exp of { alts }                    (case expression)
+      | do { stmts }                            (do expression)
+      | s"   interpolatedString "               (interpolated strings)
+      | s""" interpolatedStringMultiline """    (interpolated multiline strings)
+      | fexp
+
+Note that no space is allowed between the ``s`` and ``"``.
+
+The ``interpolatedString`` and ``interpolatedMultilineString`` rules are new rules defined as follows:
+
+.. code-block:: abnf
+
+  interpolatedString               → {graphic⟨\ | "   | ${⟩ | space | escape | gap} [ ${ exp } interpolatedString ]
+  interpolatedStringMultiline      → {{whitechar} interpolatedStringMultilineLine}
+  interpolatedStringMultilineLine  → {graphic⟨\ | """ | ${⟩ | space | escape | gap} [ ${ exp } interpolatedStringMultilineLine ]
+
+Note: We will always require interpolation to be of the form ``${foo}``, unlike other languages where ``$foo`` is allowed for single-variable interpolation. This simplifies the syntax and avoids the need for escaping bare ``$``. Interpolation may be avoided by escaping the dollar sign; e.g. ``s"\${foo}"`` is equivalent to ``"${foo}"``.
+
+The following expressions are lex errors, according to the above grammar:
+
+* ``s"a ${} b"``
+
+  * Expression is missing
+
+* ``s"a ${=} b"``
+
+  * Not an expression
+
+* ``s"a ${let x =} b"``
+
+  * Incomplete expression
+
+* ``s"a ${{b} c"``
+
+  * The second ``{`` is not a valid character to start an expression
+
+See the "Semantics" and "Examples" sections to see examples of valid interpolated strings.
+
+Machinery
+~~~~~~~~~
+
+An interpolated string expression desugars to calls to ``fromBuilder``, ``toBuilder``, and ``interpolate``, which are defined as:
+
+::
+
+  -- | Laws:
+  --     * fromBuilder . toBuilder === id
+  --     * toBuilder . fromBuilder === id
+  class Monoid (Builder s) => Buildable s where
+    type Builder s = r | r -> s
+    toBuilder :: s -> Builder s
+    fromBuilder :: Builder s -> s
+
+  class Buildable s => Interpolate a s where
+    interpolate :: a -> Builder s
```

Comment:
So this class is basically just a class of lawless conversions to string-y things. Do we really need this so much? The alternative is to have the conversions to the string type be explicit, e.g.
```
  let textExample = s"${show name1} (age: ${show age1}) encountered ${show name2} (age: ${show age2})"
```
or similarly with `tshow :: Show a => a -> Text` function.

I appreciate that the current design allows things to more often "just work", but I think the extra cost to be explicit here is not so bad, and it also makes it very easy to change the formatting of an individual item on a case-by-case basis.

Reply:
IMO that would render proposal almost completely useless.

2. It would add a ton of noise to string interpolation, so no one would use the feature
1. It wouldnt support injection-free SqlQuery, as you need to know which SqlValue to use
3. It wouldn't support escaping html by default, while allowing explicitly marking certain strings as safe raw html

I'd rather the proposal be rejected than only allow interpolating strings